File: Question_31_40/answer_32.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# Read image
img = cv2.imread("imori.jpg").astype(np.float32)
H, W, C = img.shape

# Gray scale
gray = 0.2126 * img[..., 2] + 0.7152 * img[..., 1] + 0.0722 * img[..., 0]

#fimg = np.fft.fft2(gray)
    
# 第1象限と第3象限, 第2象限と第4象限を入れ替え
#fimg =  np.fft.fftshift(fimg)
#print(fimg)
# パワースペクトルの計算
#mag = 20*np.log(np.abs(fimg))
    
# 入力画像とスペクトル画像をグラフ描画
#plt.subplot(121)
#plt.imshow(gray, cmap = 'gray')
#plt.subplot(122)
#plt.imshow(mag, cmap = 'gray')
#plt.show()


# Fourier
U = W
V = H

X = np.zeros((V, U), dtype=np.complex)
X_r = np.zeros((V, U), dtype=np.complex)
X_i = np.zeros((V, U), dtype=np.float32)

x = np.tile(np.arange(W), (H, 1))
y = np.arange(H).repeat(W).reshape(H, -1)

for v in range(V):
    for u in range(U):
        #X_r[v, u] = np.sum(re)
        #X_i[v, u] = np.sum(img)
        theta = (-2j * np.pi * (1. * v * y / H + 1. * u * x / W))
        X[v, u] = np.sum(gray * np.exp(theta)) / (H * W)

print(X)
#X = np.sqrt(np.power(X_r, 2), np.power(X_i, 2))
X = np.abs(X.imag).astype(np.uint8)
print(X)
plt.imshow(X, cmap='gray')
plt.show()

out = np.zeros((V, U), dtype=np.float32)

for v in range(V):
    for u in range(U):
        theta = (2. * np.pi * (1. * v * y / H + 1. * u * x / W))
        out[v, u] = np.sum(X_r[v, u] * np.cos(theta) + X_i[v, u] * np.sin(theta))

out = out.astype(np.uint8)
    
# Save result
cv2.imshow("result", out)
cv2.waitKey(0)
cv2.imwrite("out.jpg", out)

File: Question_31_40/answers_cpp/answer_32.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <iostream>
#include <math.h>
#include <complex>

const int height = 128, width = 128;

struct fourier_str {
  std::complex<double> coef[height][width];
};

// RGB to Gray scale
cv::Mat BGR2GRAY(cv::Mat img){
  // prepare output
  cv::Mat out = cv::Mat::zeros(height, width, CV_8UC1);
  
  // BGR -> Gray
  for (int y = 0; y < height; y++){
    for (int x = 0; x < width; x++){
      out.at<uchar>(y, x) = (int)((float)img.at<cv::Vec3b>(y, x)[0] * 0.0722 + \
				  (float)img.at<cv::Vec3b>(y, x)[1] * 0.7152 + \
				  (float)img.at<cv::Vec3b>(y, x)[2] * 0.2126);
    }
  }
  return out;
}

// Discrete Fourier transformation
fourier_str dft(cv::Mat img, fourier_str fourier_s){
  double I;
  double theta;
  std::complex<double> val;

  for (int l = 0; l < height; l ++){
    for (int k = 0; k < width; k ++){
      val.real(0);
      val.imag(0);
      for (int y = 0; y < height; y++){
        for (int x = 0; x < width; x++){
          I = (double)img.at<uchar>(y, x);
          theta = -2 * M_PI * ((double)k * (double)x / (double)width + (double)l * (double)y / (double)height);
          val += std::complex<double>(cos(theta), sin(theta)) * I;
        }
      }
      val /= sqrt(height * width);
      fourier_s.coef[l][k] = val;
    }
  }

  return fourier_s;
}

// Inverse Discrete Fourier transformation
cv::Mat idft(cv::Mat out, fourier_str fourier_s){

  double theta;
  double g;
  std::complex<double> G;
  std::complex<double> val;

  for ( int y = 0; y < height; y ++){
    for ( int x = 0; x < width; x ++){
      val.real(0);
      val.imag(0);
      for ( int l = 0; l < height; l ++){
        for ( int k = 0; k < width; k ++){
          G = fourier_s.coef[l][k];
          theta = 2 * M_PI * ((double)k * (double)x / (double)width + (double)l * (double)y / (double)height);
          val += std::complex<double>(cos(theta), sin(theta)) * G;
        }
      }
      g = std::abs(val) / sqrt(height * width);      
      out.at<uchar>(y, x) = (uchar)g;
    }
  }

  return out;
}

// Main
int main(int argc, const char* argv[]){

  // read original image
  cv::Mat img = cv::imread("imori.jpg", cv::IMREAD_COLOR);

  // Fourier coefficient
  fourier_str fourier_s;

  // output image
  cv::Mat out = cv::Mat::zeros(height, width, CV_8UC1);

  // BGR -> Gray
  cv::Mat gray = BGR2GRAY(img);

  // DFT
  fourier_s = dft(gray, fourier_s);

  // IDFT
  out = idft(out, fourier_s);
  
  //cv::imwrite("out.jpg", out);
  cv::imshow("answer", out);
  cv::waitKey(0);
  cv::destroyAllWindows();

  return 0;
}

